File: dotnet/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.Example1/VirtualProxy.cs
namespace ProxyDesignPattern.Example1;

/// <summary>
/// VIRTUAL PROXY - Implements lazy loading for video content.
/// Video info is loaded immediately, but actual content is loaded only when streaming starts.
/// </summary>
public class VirtualProxy : IVideoService
{
    private readonly IVideoService _realService;
    private readonly Dictionary<string, Video> _loadedVideos = new();

    public VirtualProxy(IVideoService realService)
    {
        _realService = realService;
    }

    public Video? GetVideoInfo(string videoId)
    {
        // Video info is lightweight, fetch it immediately
        return _realService.GetVideoInfo(videoId);
    }

    public void StreamVideo(string videoId)
    {
        // Check if video content is already loaded
        if (_loadedVideos.ContainsKey(videoId))
        {
            Console.WriteLine($"[VirtualProxy] Video already loaded, reusing content");
            return;
        }

        // Lazy load: Only load video content when actually needed
        Console.WriteLine($"[VirtualProxy] First time access - loading video content lazily");
        _realService.StreamVideo(videoId);
        
        var video = _realService.GetVideoInfo(videoId);
        if (video != null)
        {
            _loadedVideos[videoId] = video;
        }
    }
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.Example1/src/VirtualProxy.ts
import { IVideoService } from './IVideoService';
import { Video } from './Models';

// VIRTUAL PROXY — lazy loads video content
export class VirtualProxy implements IVideoService {
  private readonly loadedVideos = new Map<string, Video>();

  constructor(private readonly realService: IVideoService) {}

  getVideoInfo(videoId: string): Video | null {
    return this.realService.getVideoInfo(videoId);
  }

  streamVideo(videoId: string): void {
    if (this.loadedVideos.has(videoId)) {
      console.log('[VirtualProxy] Video already loaded, reusing content');
      return;
    }
    console.log('[VirtualProxy] First time access - loading video content lazily');
    this.realService.streamVideo(videoId);
    const video = this.realService.getVideoInfo(videoId);
    if (video) this.loadedVideos.set(videoId, video);
  }
}
